flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -91,6 +93,8 @@
 	/** The number of pods requested, but not yet granted. */
 	private int numPendingPodRequests = 0;
 
+	private KubernetesWatch podsWatch;
```

Comment:
Thanks a lot for kicking off this discussion. When i introduce these wrappers `KubernetesPod`, `KubernetesService`, `KubernetesToleration`, `KubernetesWatch` and `PodCallbackHandler`, i am just to limit the fabric8 specific classes in the package `org.apache.flink.kubernetes.kubeclient`. So then when we migrate to other kubernetes-client implementation, we will not have to change too many classes that have concrete logics. Another benefit is we could add some specific method for the wrapper resource(e.g. KubernetesPod#isTerminated).

It is a pity that we still have few fabric8 classes in `KubernetesUtils`. But i think it could be easily factored out if we want.

It is just my personal thoughts and i could rethink it if some others also have the same concerns.

Reply:
I think it is a good idea to limit the surface area of external dependencies as much as possible. This makes future refactorings (e.g. changing the actual implementation to use a different client) much easier.

Reply:
Maybe that's right. But what benefits can we gain by changing the actual implementation to use a different client?

Reply:
For example, the `fabric8` `kubernetes-client` project is no longer active in the future although it is the best choice for now.

Reply:
> For example, the fabric8 kubernetes-client project is no longer active in the future although it is the best choice for now.

Not sure why it would be no longer active?

Reply:
Every open source project has its own lifecycle. I could see the official K8s java client is becoming better with fast evolving. Maybe we will replace the actual implementation in the future. It is just personal thought. Certainly you could have your own and we do not need to reach a consensus here.

Whatever, introducing the wrapper class does not bring too much burden to our current implementation. So i will try to keep them.